### DISCUSSION TITLE
使用xmake编译出来的固件比make大，ld链接行为有差异，应该固化flash的部分没有固化

### DISCUSSION THREAD

**Discussion by PeakRacing** [xmake-io/xmake#3975] [Q&A]
![image](https://github.com/xmake-io/xmake/assets/24759100/3eb6e05d-bab5-42f7-ada7-5ccb04d2d251)

仓库地址: https://gitee.com/Dozingfiretruck/xr872_sdk/tree/xmake/ 
使用  project\code 项目进行比较
编译工具链:gcc-arm-none-eabi-4_9-2015q2

make编译方法：https://blog.csdn.net/zhoushangshi/article/details/104864754
找了好久没找到问题原因，麻烦大佬看下

**Comment by waruqi** [comment]
这种自己对下 flags 不就出来了么

**Reply by PeakRacing** [reply]
flags第一次就对过了，make的命令和xmake -v 一点一点对的，flags没错的，大佬看下.project.ld，这不是这个ld的原因导致的

**Reply by waruqi** [reply]
flags 完全一样，那就对下 project.ld 。。可以两边的命令，自己 copy 出来 ，手动执行下，脱离 xmake / makefile 。。挨个对。。

然后两边的 .o 也可以对下，大小，flags 。。不光是 链接阶段

我暂时没环境和时间测试这个

**Comment by PeakRacing** [comment]
project.ld和rom_symbol.ld预编译之后也一抹一眼，我在想是不是project.ld中这个写法影响了 
![image](https://github.com/xmake-io/xmake/assets/24759100/3cb41961-7df9-4dd8-9bc6-68aa1db2d8c2)

**Reply by waruqi** [reply]
.o 大小呢，也一样?  如果全完全一样，不可能有差异的。。

**Reply by waruqi** [reply]
> 我在想是不是project.ld中这个写法影响了

既然完全一样，那写法不也一样么。。为啥 make 就行

**Reply by PeakRacing** [reply]
https://user-images.githubusercontent.com/24759100/254455303-3eb6e05d-bab5-42f7-ada7-5ccb04d2d251.png
分析看大小一样但是连接到的位置不一样，make能正常连接到固化位置，xmake大部分没有固化

**Reply by PeakRacing** [reply]
所以我想会不会这个写法导致的？make会把所有的.o直接放在源码同目录下，xmake更干净一些，.o不在同源码目录

**Reply by waruqi** [reply]
这个应该不影响吧，，你可以一点点还原到跟 make 完全一致，.o 路径也可以改过去保持一致

或者搞个小点的工程，复现后，再两者对比测试

**Reply by PeakRacing** [reply]
确实这个会有影响，有两个地方影响了，1是xmake会生成.c.o这种的，和ld不符合，另外 *project/common/cmd*.o 这种指定路径的也会有影响 我手动吧.o改成.c.o 然后吧*project/common/cmd*.o改成*cmd*.o之后能解决固化问题，大佬看看有什么方便一些的方法能兼容？

**Reply by PeakRacing** [reply]
我把 \project\linker_script\gcc\appos.ld 修改为 \project\linker_script\gcc\appos_xmake.ld 之后固化问题基本解决了，但是如果不修改ld 修改xmake应该如何实现呢

**Reply by waruqi** [reply]
这个你只能自己改 ld 来处理，.o 路径是改不了的，.c.o 是为了避免 c/cpp 同名文件冲突，分分目录存储也是避免冲突 没法改。。

**Reply by waruqi** [reply]
你只能 set_objectdir 改总 obj 根目录

**Reply by waruqi** [reply]
或者你写个 rule 自己重写 on_link ，然后自己 os.cp 将 target:objectfiles 移动到其他地方 改名，然后自己调用 linker
